A47-xds-federation.md

```diff
@@ -0,0 +1,682 @@
+`server_listener_resource_name_template` described in [gRFC A36][A36]:
+- If the template starts with `xdstp:`, it will be interpreted as a
+  new-style name, in which case the authority of the URI will be used to
+  select the relevant configuration in the `authorities` map.
+- When replacing the `%s` token, if the template starts with `xdstp:`,
+  the replacement string will be percent-encoded in the resulting resource
+  name.
+
+The resulting field will have the following format:
+
+```
+// A template for the name of the Listener resource to subscribe to
+// for a gRPC server.
+//
+// If starts with "xdstp:", will be interpreted as a new-style name,
+// in which case the authority of the URI will be used to select the
+// relevant configuration in the "authorities" map.
+//
+// The token "%s", if present in this string, will be replaced with
+// the IP and port on which the server is listening.  If the template
+// starts with "xdstp:", the replaced string will be percent-encoded.
+//
+// There is no default; if unset, xDS-based server creation fails.
+//
+// Example:
+// "xdstp://xds.authority.com/envoy.config.listener.v3.Listener/grpc/server/%s"
+"server_listener_resource_name_template": <string>,
+```
+
+#### Example: Bootstrap Config Contains No New Fields
+
+If the bootstrap config does not contain any of the new fields described
+above, then here's how the resource names will be determined.
+
+If a gRPC client channel is created for `xds:server.example.com`:
+- The target URI has no authority, so gRPC will look at the bootstrap
+  config's `client_default_listener_resource_name_template` field, which
+  is not set, so it defaults to `%s`.  So the Listener resource name will be
+  `server.example.com`, just like it would be prior to this design.
+- The Listener resource name does not start with `xdstp:`, so we do not
+  use the `authorities` map at all.
+- We use the top-level `xds_server` list, which is the same one that we
+  would have used prior to this design.
+
+If a gRPC client channel is created for `xds://xds.authority.com/server.example.com`:
+- The `authorities` map has no entry for `xds.authority.com`, so we
+  handle this as an invalid URI, which is exactly the same way it
+  would have been handled prior to this design.
+
+If a gRPC server is created listening on `0.0.0.0:8080`:
+- The existing bootstrap config must already set the server listener
+  resource name template to some non-`xdstp` value, so we'll continue
+  to use that as the resource name.
+- The resource name from template does not start with `xdstp:`, so we
+  do not do any lookup in the `authorities` map.
+- We default to the top-level `xds_server` list, which is the same one
+  that we would have used prior to this design.
+
+#### Example: New-Style Names on gRPC Client
+
+The following new fields are added to the bootstrap config:
+
+```
+"client_default_listener_resource_name_template": "xdstp://xds.authority.com/envoy.config.listener.v3.Listener/%s",
```

Comment:
Do we check the type (`envoy.config.listener.v3.Listener` in this example) to make sure it's valid?
What if the users set this to `envoy.config.listener.v3.Cluster`? (So it's an LDS watch for a CDS resource?)
Do we fail? Or do we sent it as a LDS request, assuming the users know what they are doing?

---

Similarly, what if in an LDS response, the resource name's type is not Listener?
Do we NACK it?
Or we accept it and call the watcher?

Reply:
Good questions.  Let me answer the more general issue first, then I'll come back to the bootstrap file.

We should check the type on input, from both the `XdsClient` API and on the wire, and reject new-style resource names whose types do not match the type of resource we're told to look at:
- If a caller of the `XdsClient` API starts a watch on a resource of type (e.g.) Listener with a new-style name that indicates a different type, we should reject it.  In C-core, we're handling this by basically immediately reporting an error to the watcher and then never sending another update to it.  We don't actually subscribe to the resource on the wire.
- If we receive a resource on the wire in a response for resource type (e.g.) Listener that contains a new-name that indicates a different type, we NACK.  But that shouldn't actually happen in practice, since we should never subscribe to such a name in the first place.

Now, coming back to the question of the bootstrap file, I can think of arguments both for and against doing this validation.  Here are some of the factors to consider:
- The expansion of this template will be done outside of the `XdsClient` (on the client side, it will be done in the `xds` resolver; on the server side, it will be done in the xDS-specific server code), and the caller will then start a watch for the resulting resource name on the `XdsClient`.  So technically, we don't really need to do this validation at bootstrap-reading time, because the watcher will simply fail when we start a watch on the resulting resource name.  (Counter-argument: Surfacing the error earlier might be more friendly to users, if it turns out that this is a common mistake.)
- There may be cases in the future where it will be legitimate to have a different resource type here.  For example, we might someday want to be able to support pointing to an individual entry within a ListenerCollection resource (see xRFC TP1 for info on [list collection resources](https://github.com/cncf/xds/blob/main/proposals/TP1-xds-transport-next.md#list) and the [`#entry=<name>` fragment syntax for xdstp URIs](https://github.com/cncf/xds/blob/main/proposals/TP1-xds-transport-next.md#uri-based-xds-resource-names)).  (Counter-argument: If/when we ever decide to support this, we could change the validation we do at bootstrap parsing time.  But having older clients flat-out reject this might cause more of a problem for backward-compatibility.)
- In principal, it should be possible for the `%s` token to include the resource type, in which case we can't actually do this validation in the first place.  For example, it should be possible for the template to be `xdstp:%s` and for the client to create a channel for `xds:envoy.config.listener.v3.Listener/server.example.com`.  Or, there could be an even more esoteric case, like the template being `xdstp:envoy.config.listener.v3.%s` and the channel being created for `xds:Listener/server.example.com`.  (Counter-argument: I can't think of a reason for anyone to actually do this... although there could potentially be such a case in the future, if/when we support ListenerCollection resources.)

There's enough of a grey area here that my preference would be to say that we should *not* explicitly check this at bootstrap parsing time, but I'm open to being convinced otherwise if you or other reviewers feel otherwise.  WDYT?

Reply:
Not checking at bootstrap parsing time SGTM.

One note is that, even if we don't check at bootstrap parsing time, but we check at `client.Watch` time, it won't support any new types. So this decision doesn't affect backward-compability.

Not checking at bootstrap parsing time also has the advantage that it will fail only if the invalid template string is used. If it's set for an authority that's never picked, it won't cause any errors. (Or maybe this is a disadvantage?)

Reply:
Good point.  IMHO, the fact that any given template could affect only a subset of usage and might never actually be used in a given client is another good reason not to do this validation at bootstrap parsing time, since that does provide more options for forward-compatibility.

Reply:
> the fact that any given template could affect only a subset of usage and might never actually be used in a given client is another good reason not to do this validation at bootstrap parsing time, since that does provide more options for forward-compatibility.

Delayed validation has no relation to forward-compatibility. Since it _may_ be used by a client, you can't break the older clients with new configuration. So it seems safe to validate up-front. This also has the benefit that someone testing the bootstrap doesn't need to test all client code paths in hopes of finding every last delayed validation. In short: if it would break at runtime and we need to validate it anyway, then validating it as early as possible is beneficial.

Reply:
I don't think this is completely divorced from forward compatibility.  I think there could be cases where we would want to use a new feature for a particular new authority but not break any existing authorities.  In that case, validating this at bootstrap parsing time would make that impossible, whereas defering the check to run-time would provide the flexibility to say that old clients won't work with the new authority but will continue to work with existing authorities.

That having been said, although I still lean toward not doing this validation at bootstrap-parsing time, I can still see both sides here.  Does anyone else have any strong opinions on this?

Reply:
> I think there could be cases where we would want to use a new feature for a particular new authority but not break any existing authorities.

I see. I buy that to a degree. I do wonder "what would that feature look like" given how we've been maintaining compatibility to date by adding new fields, so that simply shouldn't be a problem for the majority case. But I still agree there may be something there for some case. This specific case, however, is a new field so that argument doesn't really hold.

Looking again at this case, I'm noticing now we'd have to parse the URL before the `%s` replacement, which seems bad. So I'm now agreeing we shouldn't validate this early. Yes, we could replace empty string or some such, but it won't be reliable.

Reply:
That's a good point.  Sounds like we've converged on this.